cqrs-postgres/src/reactor.rs

```diff
@@ -0,0 +1,633 @@
+
+            let raw_events = conn
+                .read_all_events(&query_with_args, since, params.as_slice())
+                .map_err(DbError::postgres)?;
+
+            for event in raw_events {
+                let event_id = event.event_id;
+                reaction.react(event).map_err(DbError::react)?;
+
+                conn.save_since(R::reaction_name(), event_id)
+                    .map_err(DbError::postgres)?;
+
+                event_count += 1;
+            }
+
+            drop(conn);
+
+            ::std::thread::sleep(R::interval());
+        }
+
+        Ok(event_count)
+    }
+
+    fn generate_query_with_args(
+        &self,
+        predicate: ReactionPredicate,
+        params: &mut Vec<Box<dyn ToSql>>,
+        max_count: u64,
+    ) -> Result<String, DbError> {
+        let max_count = Box::new(max_count.min(i64::max_value() as u64) as i64);
+
+        match predicate.aggregate_predicate {
+            AggregatePredicate::AllAggregates(EventTypesPredicate::AllEventTypes) => {
+                params.push(max_count);
+
+                Ok(String::from(
+                    "SELECT event_id, aggregate_type, entity_id, sequence, event_type, payload \
+                     FROM events \
+                     WHERE event_id > $1 \
+                     ORDER BY event_id ASC \
+                     LIMIT $2",
+                ))
+            }
+            AggregatePredicate::AllAggregates(EventTypesPredicate::SpecificEventTypes(
+                event_types,
+            )) => {
+                params.push(Box::new(event_types));
+                params.push(max_count);
+
+                Ok(String::from(
+                    "SELECT event_id, aggregate_type, entity_id, sequence, event_type, payload \
+                     FROM events \
+                     WHERE event_id > $1 \
+                     AND event_type = ANY ($2) \
+                     ORDER BY event_id ASC \
+                     LIMIT $3",
+                ))
+            }
+            AggregatePredicate::SpecificAggregates(aggregate_predicates) => {
+                let mut query = String::from(
+                    "SELECT event_id, aggregate_type, entity_id, sequence, event_type, payload \
+                     FROM events \
+                     WHERE event_id > $1 AND (FALSE",
+                );
+
+                let mut param_count = 1;
+
+                for predicate in aggregate_predicates {
+                    match &predicate.event_types {
+                        EventTypesPredicate::SpecificEventTypes(event_types) => {
+                            write!(
+                                query,
+                                " OR (aggregate_type = ${} AND event_type = ANY (${}))",
+                                param_count + 1,
+                                param_count + 2
+                            )
+                            .map_err(DbError::format)?;
```

Comment:
We could eliminate the `Format` arm all together here. We are writing to a string, which will only fail for out-of-memory, and we're formatting integers, which also won't fail. We can `unwrap()` or `expect("formatting integers into a string never fails")` here.

Reply:
Done